src/native/clr/include/host/pinvoke-override-impl.hh

```diff
@@ -0,0 +1,177 @@
+#endif
+
+#include <mutex>
+#include <string>
+
+#include "pinvoke-override.hh"
+#include "../runtime-base/logger.hh"
+#include "../runtime-base/monodroid-dl.hh"
+#include "../runtime-base/startup-aware-lock.hh"
+
+namespace xamarin::android {
+	PINVOKE_OVERRIDE_INLINE
+	auto PinvokeOverride::load_library_symbol (std::string_view const& library_name, std::string_view const& symbol_name, void **dso_handle) noexcept -> void*
+	{
+		void *lib_handle = dso_handle == nullptr ? nullptr : *dso_handle;
+
+		if (lib_handle == nullptr) {
+			// We're being called as part of the p/invoke mechanism, we don't need to look in the AOT cache
+			constexpr bool PREFER_AOT_CACHE = false;
+
+			// Handle p/invokes of the form [DllImport ("liblog")] or [DllImport ("log")]
+			// TODO: try modifying the name to contain both the `log` prefix and the `.so` suffix
+			dynamic_local_path_string short_library_name;
+			if (!Util::path_has_directory_components (library_name)) {
+				if (!library_name.starts_with (Constants::DSO_PREFIX)) {
+					short_library_name.append (Constants::DSO_PREFIX);
+				}
+				short_library_name.append (library_name);
+				if (!short_library_name.ends_with (Constants::dso_suffix)) {
+					short_library_name.append (Constants::dso_suffix);
+				}
+
+				log_debug (LOG_ASSEMBLY, "Modified p/invoke library name to '{}'", short_library_name.get ());
+				lib_handle = MonodroidDl::monodroid_dlopen<PREFER_AOT_CACHE> (short_library_name.get (), microsoft::java_interop::JAVA_INTEROP_LIB_LOAD_LOCALLY);
+			}
+
+			if (lib_handle == nullptr) {
+				lib_handle = MonodroidDl::monodroid_dlopen<PREFER_AOT_CACHE> (library_name, microsoft::java_interop::JAVA_INTEROP_LIB_LOAD_LOCALLY);
+			}
+
+			if (lib_handle == nullptr) {
+				log_warn (LOG_ASSEMBLY, "Shared library '{}' not loaded, p/invoke '{}' may fail", library_name, symbol_name);
+				return nullptr;
+			}
+
+			if (dso_handle != nullptr) {
+				void *expected_null = nullptr;
+				if (!__atomic_compare_exchange (dso_handle, &expected_null, &lib_handle, false /* weak */, __ATOMIC_ACQUIRE /* success_memorder */, __ATOMIC_RELAXED /* xxxfailure_memorder */)) {
+					log_debug (LOG_ASSEMBLY, "Library '{}' handle already cached by another thread", library_name);
+				}
+			}
+		}
+
+		void *entry_handle = MonodroidDl::monodroid_dlsym (lib_handle, symbol_name);
+		if (entry_handle == nullptr) {
+			log_warn (LOG_ASSEMBLY, "Symbol '{}' not found in shared library '{}', p/invoke may fail", library_name, symbol_name);
+			return nullptr;
+		}
+
+		return entry_handle;
+	}
+
+	PINVOKE_OVERRIDE_INLINE
+	auto PinvokeOverride::load_library_entry (std::string const& library_name, std::string const& entrypoint_name, pinvoke_api_map_ptr api_map) noexcept -> void*
+	{
+		// Make sure some other thread hasn't just added the entry
+		auto iter = api_map->find (entrypoint_name);
+		if (iter != api_map->end () && iter->second != nullptr) {
+			return iter->second;
+		}
```

Comment:
Why not locking the load_library_entry function? I noticed the pinvoke_map_write_lock why not doing it here?

Reply:
It is locked only when necessary (see e.g. the `fetch_or_create_pinvoke_map_entry` function), to minimize overhead (especially during startup before we branch out to the managed land, where we know there's just one thread running)